text/0000-gen-fn.md

```diff
@@ -0,0 +1,785 @@
+document, but we do not have to settle them with this RFC. We'll describe
+current thinking on the semantics, but some questions will be left open to be
+answered at a later time after we gain more experience with the
+implementation.
+
+Writing iterators manually can be very painful. Many iterators can be written by
+chaining `Iterator` methods, but some need to be written as a `struct` and have
+`Iterator` implemented for them. Some of the code that is written this way
+pushes people to avoid iterators and instead execute a `for` loop that eagerly
+writes values to mutable state. With this RFC, one can write the `for` loop
+and still get a lazy iterator of values.
+
+As an example, here are multiple ways to write an iterator over something that contains integers
+while only keeping the odd integers and multiplying each by 2:
+
+```rust
+// `Iterator` methods
+fn odd_dup(values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
+    values.filter(|value| value.is_odd()).map(|value| value * 2)
+}
+
+// `std::iter::from_fn`
+fn odd_dup(mut values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
+    std::iter::from_fn(move || {
+        loop {
+            let value = values.next()?;
+            if value % 2 == 1 {
+                return Some(value * 2);
+            }
+        }
+    })
+}
+
+// `struct` and manual `impl`
+fn odd_dup(values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
+    struct Foo<T>(T);
+    impl<T: Iterator<Item = u32>> Iterator<Item = u32> for Foo<T> {
+        type Item = u32;
+        fn next(&mut self) -> Option<u32> {
+            loop {
+                let value = self.0.next()?;
+                if value.is_odd() {
+                    return Some(x * 2)
+                }
+            }
+        }
+    }
+    Foo(values)
+}
+
+// `gen block`
+fn odd_dup(values: impl Iterator<Item = u32>) -> impl Iterator<Item = u32> {
+    gen {
+        for value in values {
+            if value.is_odd() {
+                yield value * 2;
+            }
+        }
+    }
+}
+
+// `gen fn`
+gen fn odd_dup(values: impl Iterator<Item = u32>) -> u32 {
+    for value in values {
+        if value.is_odd() {
+            yield value * 2;
+        }
+    }
+}
+```
+
+Iterators created with `gen` return `None` once they `return` (implicitly at the end of the scope or explicitly with `return`).
+See [the unresolved questions][unresolved-questions] for whether `gen` iterators are fused or may behave strangely after having returned `None` once.
+Under no circumstances will it be undefined behavior if `next` is invoked again after having gotten a `None`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## New keyword
+
+Starting in the 2024 edition, `gen` is a keyword that cannot be used for naming any items or bindings. This means during the migration to the 2024 edition, all variables, functions, modules, types, etc. named `gen` must be renamed.
```

Comment:
```suggestion
Starting in the 2024 edition, `gen` is a keyword that cannot be used for naming any items or bindings. This means during the migration to the 2024 edition, all variables, functions, modules, types, etc. named `gen` must be renamed or migrated to raw identifier syntax (`r#gen`).
```

Some crates may be unable to rename items in their public API.

Reply:
We've since updated the document to cover this.  Thanks for the note.